backstage/community-plugins
# feat: add msgraph catalog transformer module

### Description
The default user transformer provided by the [MsGraph plugin](https://github.com/backstage/backstage/tree/918d273bbb843bd09bb6c80d22d87c257a7e5343/plugins/catalog-backend-module-msgraph#readme) does not ingest users who do not have an email included in their profile, preventing those users from signing in. This is an issue because email is not a mandatory field when creating an user and therefore not all users will have this field defined.

This module [customizes the entity provider](https://github.com/backstage/backstage/tree/918d273bbb843bd09bb6c80d22d87c257a7e5343/plugins/catalog-backend-module-msgraph#customize-the-processor-or-entity-provider) by implementing the `microsoftGraphOrgEntityProviderTransformExtensionPoint` to allow users without an email to be ingested into the catalog by using `userPrincipalName` as the identifier, instead of email by default.
* The `userPrincipalName` parameter is must exist for each user and is unique.

Note: a PR was originally opened [here](https://github.com/janus-idp/backstage-showcase/pull/1440) as an internal plugin to the backstage-showcase instance but in order to support dynamic installation of this module, we would need to publish it here. 

#### Fixes
Fixes of [RHIDP-3652](https://issues.redhat.com/browse/RHIDP-3652)

#### To test
1. Add an Microsoft Entra ID user without email
2. [Set up](https://github.com/backstage/backstage/tree/918d273bbb843bd09bb6c80d22d87c257a7e5343/plugins/catalog-backend-module-msgraph#getting-started) the `microsoftGraphOrg` entity provider as normal.
3. Verify that they are ingested into the catalog and that the entityRef is the `userPrincipalName`

#### :heavy_check_mark: Checklist
